# DATA PRE-PROCESSING ASSIGNMENT

# Team Member
#### 1) Avi Bomb - 106117017
#### 2) Charan Kumar - 106117019
#### 3) Dhiraj Kumar - 106117024
#### 4) Suraksha - 106117075
#### 5) Vidhya Lakshmi B - 106117107

# Problem Statement & Solution

### These statments are for reading in data from a CSV file using Pandas CSV read. The given dataset has 42553 Rows and 115 Columns initially. It has NULL values, Duplicate rows, Redundant Columns, Categorical values and Unique value columns. After applying the necessary preprocessing steps, dataset size got reduced and it got rid of all the problems mentioned above. The final dataset at the end of this documentation is fit for applying machine learning algorithms efficiently.

In [1]:
import pandas as pd
dataset = pd.read_csv('lending-club-loans.csv', low_memory=False)

In [2]:
print(dataset)

                                                    id  member_id  loan_amnt  \
0                                              1077501  1296599.0     5000.0   
1                                              1077430  1314167.0     2500.0   
2                                              1077175  1313524.0     2400.0   
3                                              1076863  1277178.0    10000.0   
4                                              1075358  1311748.0     3000.0   
...                                                ...        ...        ...   
42548                                           505136   650770.0    20000.0   
42549                                           473367   598172.0     1750.0   
42550                                            70686    70681.0     5000.0   
42551  Total amount funded in policy code 1: 460296150        NaN        NaN   
42552          Total amount funded in policy code 2: 0        NaN        NaN   

       funded_amnt  funded_amnt_inv    

### Checking for the existance of Duplicate Rows.

In [3]:
from sklearn import preprocessing
import matplotlib.pyplot as plt
import seaborn as sns
any(dataset.duplicated())

True

### Since duplicate rows are present the following code keeps the first occurance and drops the remaining duplicate rows based on 'id' attribute. After the removal of duplicate rows, we have 42538 Rows and 115 columns in the  dataset.

In [4]:
unique_rows=dataset.drop_duplicates(subset=['id'], keep='first')

In [5]:
print(unique_rows)

                                                    id  member_id  loan_amnt  \
0                                              1077501  1296599.0     5000.0   
1                                              1077430  1314167.0     2500.0   
2                                              1077175  1313524.0     2400.0   
3                                              1076863  1277178.0    10000.0   
4                                              1075358  1311748.0     3000.0   
...                                                ...        ...        ...   
42534                                            72176    70868.0     2525.0   
42535                                            71623    70735.0     6500.0   
42550                                            70686    70681.0     5000.0   
42551  Total amount funded in policy code 1: 460296150        NaN        NaN   
42552          Total amount funded in policy code 2: 0        NaN        NaN   

       funded_amnt  funded_amnt_inv    

### Finding the number of NULL values in each Row and Column. Axis value for Columns and Rows is '0' and '1' respectively.

In [6]:
unique_rows.isnull().sum(axis = 1)

0         58
1         57
2         59
3         56
4         55
        ... 
42534     70
42535     71
42550     69
42551    114
42552    114
Length: 42538, dtype: int64

In [7]:
unique_rows.isnull().sum(axis = 0)

id                                0
member_id                         3
loan_amnt                         3
funded_amnt                       3
funded_amnt_inv                   3
                              ...  
tax_liens                       108
tot_hi_cred_lim               42538
total_bal_ex_mort             42538
total_bc_limit                42538
total_il_high_credit_limit    42538
Length: 115, dtype: int64

### Checking and dropping columns with more than half NULL values. After dropping columns with more than half NULL values, we have 42538 Rows and 58 Columns.

In [8]:
half_count_rows = len(unique_rows) / 2

In [9]:
print(half_count_rows)

21269.0


In [10]:
less_null = unique_rows.dropna(thresh=half_count_rows,axis=1)

In [11]:
print(less_null)

                                                    id  member_id  loan_amnt  \
0                                              1077501  1296599.0     5000.0   
1                                              1077430  1314167.0     2500.0   
2                                              1077175  1313524.0     2400.0   
3                                              1076863  1277178.0    10000.0   
4                                              1075358  1311748.0     3000.0   
...                                                ...        ...        ...   
42534                                            72176    70868.0     2525.0   
42535                                            71623    70735.0     6500.0   
42550                                            70686    70681.0     5000.0   
42551  Total amount funded in policy code 1: 460296150        NaN        NaN   
42552          Total amount funded in policy code 2: 0        NaN        NaN   

       funded_amnt  funded_amnt_inv    

### The following output we can see that the total number of NULL values in each column is lesser than Half the number of Rows in the data set.

In [12]:
less_null.isnull().sum(axis = 0)

id                                0
member_id                         3
loan_amnt                         3
funded_amnt                       3
funded_amnt_inv                   3
term                              3
int_rate                          3
installment                       3
grade                             3
sub_grade                         3
emp_title                      2629
emp_length                     1115
home_ownership                    3
annual_inc                        7
verification_status               3
issue_d                           3
loan_status                       3
pymnt_plan                        3
url                               3
desc                          13296
purpose                           3
title                            16
zip_code                          3
addr_state                        3
dti                               3
delinq_2yrs                      32
earliest_cr_line                 32
fico_range_low              

### Removing columns having only one value throughout. After dropping columns with single value throughout, we have 42538 Rows and 53 Columns.

In [13]:
same_val_col_removed = less_null.loc[:,less_null.apply(pd.Series.nunique) != 1]

In [14]:
print(same_val_col_removed)

                                                    id  member_id  loan_amnt  \
0                                              1077501  1296599.0     5000.0   
1                                              1077430  1314167.0     2500.0   
2                                              1077175  1313524.0     2400.0   
3                                              1076863  1277178.0    10000.0   
4                                              1075358  1311748.0     3000.0   
...                                                ...        ...        ...   
42534                                            72176    70868.0     2525.0   
42535                                            71623    70735.0     6500.0   
42550                                            70686    70681.0     5000.0   
42551  Total amount funded in policy code 1: 460296150        NaN        NaN   
42552          Total amount funded in policy code 2: 0        NaN        NaN   

       funded_amnt  funded_amnt_inv    

### Loading the dataset description file and renaming the columns.

In [15]:
data_dictionary = pd.read_csv('LCDataDictionary.csv', low_memory=False)
data_dictionary = data_dictionary.rename(columns={'LoanStatNew': 'name'})

In [16]:
print(data_dictionary)

                      name                                        Description
0           acc_now_delinq  The number of accounts on which the borrower i...
1     acc_open_past_24mths         Number of trades opened in past 24 months.
2               addr_state  The state provided by the borrower in the loan...
3                 all_util              Balance to credit limit on all trades
4               annual_inc  The self-reported annual income provided by th...
..                     ...                                                ...
112    verification_status  Indicates if income was verified by LC, not ve...
113  verified_status_joint  Indicates if the co-borrowers' joint income wa...
114               zip_code  The first 3 numbers of the zip code provided b...
115                    NaN                                                NaN
116                    NaN  * Employer Title replaces Employer Name for al...

[117 rows x 2 columns]


### Merging 2 Data Frames 'data_dictionary' and 'same_val_col_removed' to create a new Data Frame to help identify redundant columns.

In [17]:
combine = pd.DataFrame(same_val_col_removed.dtypes,columns=['dtypes'])
combine = combine.reset_index()
combine['name'] = combine['index']
combine = combine[['name','dtypes']]
combine['first value'] = same_val_col_removed.loc[0].values
preview = combine.merge(data_dictionary, on='name',how='left')

In [18]:
print(preview)

                       name   dtypes  \
0                        id   object   
1                 member_id  float64   
2                 loan_amnt  float64   
3               funded_amnt  float64   
4           funded_amnt_inv  float64   
5                      term   object   
6                  int_rate   object   
7               installment  float64   
8                     grade   object   
9                 sub_grade   object   
10                emp_title   object   
11               emp_length   object   
12           home_ownership   object   
13               annual_inc  float64   
14      verification_status   object   
15                  issue_d   object   
16              loan_status   object   
17               pymnt_plan   object   
18                      url   object   
19                     desc   object   
20                  purpose   object   
21                    title   object   
22                 zip_code   object   
23               addr_state   object   


In [19]:
same_val_col_removed.columns.tolist()

['id',
 'member_id',
 'loan_amnt',
 'funded_amnt',
 'funded_amnt_inv',
 'term',
 'int_rate',
 'installment',
 'grade',
 'sub_grade',
 'emp_title',
 'emp_length',
 'home_ownership',
 'annual_inc',
 'verification_status',
 'issue_d',
 'loan_status',
 'pymnt_plan',
 'url',
 'desc',
 'purpose',
 'title',
 'zip_code',
 'addr_state',
 'dti',
 'delinq_2yrs',
 'earliest_cr_line',
 'fico_range_low',
 'fico_range_high',
 'inq_last_6mths',
 'open_acc',
 'pub_rec',
 'revol_bal',
 'revol_util',
 'total_acc',
 'out_prncp',
 'out_prncp_inv',
 'total_pymnt',
 'total_pymnt_inv',
 'total_rec_prncp',
 'total_rec_int',
 'total_rec_late_fee',
 'recoveries',
 'collection_recovery_fee',
 'last_pymnt_d',
 'last_pymnt_amnt',
 'last_credit_pull_d',
 'last_fico_range_high',
 'last_fico_range_low',
 'acc_now_delinq',
 'delinq_amnt',
 'pub_rec_bankruptcies',
 'tax_liens']

### Identifying and dropping the redundant and unnecessary columns. After removing the redundant and unnecessary columns we have, we have 42538 Rows and 31 Columns in the dataset.

In [20]:
drop_redundant_coloumns=['url','desc','id','member_id','funded_amnt','funded_amnt_inv','int_rate','sub_grade','emp_title','issue_d','zip_code','out_prncp','out_prncp_inv','total_pymnt','total_pymnt_inv','total_rec_prncp','total_rec_int','total_rec_late_fee','recoveries','collection_recovery_fee','last_pymnt_d','last_pymnt_amnt']
print(len(drop_redundant_coloumns))

22


In [21]:
redundant_col_removed = same_val_col_removed.drop(drop_redundant_coloumns,axis=1)
redundant_col_removed.columns.tolist()

['loan_amnt',
 'term',
 'installment',
 'grade',
 'emp_length',
 'home_ownership',
 'annual_inc',
 'verification_status',
 'loan_status',
 'pymnt_plan',
 'purpose',
 'title',
 'addr_state',
 'dti',
 'delinq_2yrs',
 'earliest_cr_line',
 'fico_range_low',
 'fico_range_high',
 'inq_last_6mths',
 'open_acc',
 'pub_rec',
 'revol_bal',
 'revol_util',
 'total_acc',
 'last_credit_pull_d',
 'last_fico_range_high',
 'last_fico_range_low',
 'acc_now_delinq',
 'delinq_amnt',
 'pub_rec_bankruptcies',
 'tax_liens']

In [22]:
print(redundant_col_removed)

       loan_amnt        term  installment grade emp_length home_ownership  \
0         5000.0   36 months       162.87     B  10+ years           RENT   
1         2500.0   60 months        59.83     C   < 1 year           RENT   
2         2400.0   36 months        84.33     C  10+ years           RENT   
3        10000.0   36 months       339.31     C  10+ years           RENT   
4         3000.0   60 months        67.79     B     1 year           RENT   
...          ...         ...          ...   ...        ...            ...   
42534     2525.0   36 months        80.69     B   < 1 year           RENT   
42535     6500.0   36 months       204.84     A   < 1 year           NONE   
42550     5000.0   36 months       156.11     A  10+ years       MORTGAGE   
42551        NaN         NaN          NaN   NaN        NaN            NaN   
42552        NaN         NaN          NaN   NaN        NaN            NaN   

       annual_inc verification_status  \
0         24000.0            Verif

### Checking and dropping rows with more than half NULL values. After dropping rows with more than half NULL values, we have 42535 Rows and 31 Columns in the dataset.

In [23]:
half_count_cols = redundant_col_removed.shape[1] / 2

In [24]:
print(half_count_cols)

15.5


In [25]:
null_rows_removed = redundant_col_removed.dropna(thresh=half_count_cols,axis=0)

In [26]:
print(null_rows_removed)

       loan_amnt        term  installment grade emp_length home_ownership  \
0         5000.0   36 months       162.87     B  10+ years           RENT   
1         2500.0   60 months        59.83     C   < 1 year           RENT   
2         2400.0   36 months        84.33     C  10+ years           RENT   
3        10000.0   36 months       339.31     C  10+ years           RENT   
4         3000.0   60 months        67.79     B     1 year           RENT   
...          ...         ...          ...   ...        ...            ...   
42532     3500.0   36 months       113.39     C   < 1 year           RENT   
42533     1000.0   36 months        32.11     B   < 1 year           RENT   
42534     2525.0   36 months        80.69     B   < 1 year           RENT   
42535     6500.0   36 months       204.84     A   < 1 year           NONE   
42550     5000.0   36 months       156.11     A  10+ years       MORTGAGE   

       annual_inc verification_status  \
0         24000.0            Verif

### Finding data types of all columns to find the the existance of Typos, Categorical Values and Mixed Data Types in our dataset wherever the data type is 'object'.

In [27]:
null_rows_removed.dtypes

loan_amnt               float64
term                     object
installment             float64
grade                    object
emp_length               object
home_ownership           object
annual_inc              float64
verification_status      object
loan_status              object
pymnt_plan               object
purpose                  object
title                    object
addr_state               object
dti                     float64
delinq_2yrs             float64
earliest_cr_line         object
fico_range_low          float64
fico_range_high         float64
inq_last_6mths          float64
open_acc                float64
pub_rec                 float64
revol_bal               float64
revol_util              float64
total_acc               float64
last_credit_pull_d       object
last_fico_range_high    float64
last_fico_range_low     float64
acc_now_delinq           object
delinq_amnt             float64
pub_rec_bankruptcies    float64
tax_liens                object
dtype: o

### Replacing categorical values for columns 'emp_length', 'grade','home_ownership', 'verification_status', 'purpose' and 'addr_state' as shown in the output.

In [28]:
corelation_dict = {"emp_length": {"10+ years": 10,"9 years": 9,"8 years": 8,"7 years": 7,"6 years":
6, "5 years": 5,"4 years": 4,"3 years": 3,"2 years": 2,"1 year": 1,"< 1 year": 0,"n/a":
0},"grade":{"A": 1,"B": 2,"C": 3,"D": 4,"E": 5,"F": 6,"G": 7}}
categorical_handle = null_rows_removed.replace(corelation_dict)

In [29]:
print(categorical_handle)

       loan_amnt        term  installment  grade  emp_length home_ownership  \
0         5000.0   36 months       162.87      2        10.0           RENT   
1         2500.0   60 months        59.83      3         0.0           RENT   
2         2400.0   36 months        84.33      3        10.0           RENT   
3        10000.0   36 months       339.31      3        10.0           RENT   
4         3000.0   60 months        67.79      2         1.0           RENT   
...          ...         ...          ...    ...         ...            ...   
42532     3500.0   36 months       113.39      3         0.0           RENT   
42533     1000.0   36 months        32.11      2         0.0           RENT   
42534     2525.0   36 months        80.69      2         0.0           RENT   
42535     6500.0   36 months       204.84      1         0.0           NONE   
42550     5000.0   36 months       156.11      1        10.0       MORTGAGE   

       annual_inc verification_status  \
0         

In [30]:
print (categorical_handle['home_ownership'].value_counts())

RENT        20181
MORTGAGE    18959
OWN          3251
OTHER         136
NONE            8
Name: home_ownership, dtype: int64


In [31]:
corelation_dict = {"home_ownership": {"RENT": 1,"MORTGAGE": 2,"OWN": 3,"OTHER": 7,"NONE":0}}
categorical_handle = categorical_handle.replace(corelation_dict)

In [32]:
print (categorical_handle['home_ownership'].value_counts())

1    20181
2    18959
3     3251
7      136
0        8
Name: home_ownership, dtype: int64


In [33]:
print (categorical_handle['verification_status'].value_counts())

Not Verified       18758
Verified           13471
Source Verified    10306
Name: verification_status, dtype: int64


In [34]:
corelation_dict = {"verification_status": {"Not Verified": 0,"Verified": 1,"Source Verified": 2}}
categorical_handle = categorical_handle.replace(corelation_dict)

In [35]:
print (categorical_handle['verification_status'].value_counts())

0    18758
1    13471
2    10306
Name: verification_status, dtype: int64


In [36]:
print (categorical_handle['purpose'].value_counts())

debt_consolidation    19776
credit_card            5477
other                  4425
home_improvement       3199
major_purchase         2311
small_business         1992
car                    1615
wedding                1004
medical                 753
moving                  629
house                   426
educational             422
vacation                400
renewable_energy        106
Name: purpose, dtype: int64


In [37]:
corelation_dict = {"purpose": {"debt_consolidation": 0,"credit_card": 1,"home_improvement": 2,"major_purchase": 3,"small_business": 4,"car": 5,"wedding": 6,"medical": 7,"moving": 8,"house": 9,"educational": 10,"vacation": 11,"renewable_energy": 12,"other": 13}}
categorical_handle = categorical_handle.replace(corelation_dict)

In [38]:
print (categorical_handle['purpose'].value_counts())

0     19776
1      5477
13     4425
2      3199
3      2311
4      1992
5      1615
6      1004
7       753
8       629
9       426
10      422
11      400
12      106
Name: purpose, dtype: int64


In [39]:
print (categorical_handle['addr_state'].value_counts())

CA    7429
NY    4065
FL    3104
TX    2915
NJ    1988
IL    1672
PA    1651
GA    1503
VA    1487
MA    1438
OH    1329
MD    1125
AZ     933
WA     888
CO     857
NC     830
CT     816
MI     796
MO     765
MN     652
NV     527
WI     516
SC     489
AL     484
OR     468
LA     461
KY     359
OK     317
KS     298
UT     278
AR     261
DC     224
RI     208
NM     205
NH     188
WV     187
HI     181
DE     136
MT      96
WY      87
AK      86
SD      67
VT      57
TN      32
MS      26
IN      19
IA      12
NE      11
ID       9
ME       3
Name: addr_state, dtype: int64


In [40]:
corelation_dict = {"addr_state": {"CA": 0,"NY": 1,"FL": 2,"TX": 3,"NJ": 4,"IL": 5,"PA": 6,"GA": 7,"VA": 8,"MA": 9,"OH": 10,"MD": 12,"AZ": 13,"WA": 14,"CO": 15,"NC": 16,"CT": 17,"MI": 18,"MO": 19,"MN": 20,"NV": 21,"WI": 22,"SC": 23,"AL": 24,"OR": 25,"LA": 26,"KY": 27,"OK": 28,"KS": 29,"UT": 31,"AR": 32,"DC": 33,"RI": 34,"NM": 35,"NH": 36,"WV": 37,"HI": 38,"DE": 39,"MT": 40,"WY": 41,"AK": 42,"SD": 43,"VT": 44,"TN": 45,"MS": 46,"IN": 47,"IA": 48,"NE": 49,"ID": 30,"ME": 11}}
categorical_handle = categorical_handle.replace(corelation_dict)
print (categorical_handle['addr_state'].value_counts())

0     7429
1     4065
2     3104
3     2915
4     1988
5     1672
6     1651
7     1503
8     1487
9     1438
10    1329
12    1125
13     933
14     888
15     857
16     830
17     816
18     796
19     765
20     652
21     527
22     516
23     489
24     484
25     468
26     461
27     359
28     317
29     298
31     278
32     261
33     224
34     208
35     205
36     188
37     187
38     181
39     136
40      96
41      87
42      86
43      67
44      57
45      32
46      26
47      19
48      12
49      11
30       9
11       3
Name: addr_state, dtype: int64


# Target class :

### Replacing categorical values for 'loan_status' as shown in the output. The taget class for the dataset is 'loan_status'.

In [41]:
print (categorical_handle['loan_status'].value_counts())

Fully Paid                                             33586
Charged Off                                             5653
Does not meet the credit policy. Status:Fully Paid      1988
Does not meet the credit policy. Status:Charged Off      761
Current                                                  513
In Grace Period                                           16
Late (31-120 days)                                        12
Late (16-30 days)                                          5
Default                                                    1
Name: loan_status, dtype: int64


In [42]:
corelation_dict = {"loan_status": {"Fully Paid": 0,"Charged Off": 1,"Does not meet the credit policy. Status:Fully Paid": 2,"Does not meet the credit policy. Status:Charged Off": 3,"Current": 4,"In Grace Period": 5,"Late (31-120 days)": 6,"Late (16-30 days)": 7,"Default": 8}}
categorical_handle = categorical_handle.replace(corelation_dict)

In [43]:
print (categorical_handle['loan_status'].value_counts())

0    33586
1     5653
2     1988
3      761
4      513
5       16
6       12
7        5
8        1
Name: loan_status, dtype: int64


In [44]:
print(categorical_handle)

       loan_amnt        term  installment  grade  emp_length  home_ownership  \
0         5000.0   36 months       162.87      2        10.0               1   
1         2500.0   60 months        59.83      3         0.0               1   
2         2400.0   36 months        84.33      3        10.0               1   
3        10000.0   36 months       339.31      3        10.0               1   
4         3000.0   60 months        67.79      2         1.0               1   
...          ...         ...          ...    ...         ...             ...   
42532     3500.0   36 months       113.39      3         0.0               1   
42533     1000.0   36 months        32.11      2         0.0               1   
42534     2525.0   36 months        80.69      2         0.0               1   
42535     6500.0   36 months       204.84      1         0.0               0   
42550     5000.0   36 months       156.11      1        10.0               2   

       annual_inc  verification_status 

### Removing 'months' from 'term' column using rstrip and typecasting it to float data type as shown in the output.

In [45]:
categorical_handle['term'] = categorical_handle['term'].str.rstrip(' months').astype('float')

In [46]:
print(categorical_handle)

       loan_amnt  term  installment  grade  emp_length  home_ownership  \
0         5000.0  36.0       162.87      2        10.0               1   
1         2500.0  60.0        59.83      3         0.0               1   
2         2400.0  36.0        84.33      3        10.0               1   
3        10000.0  36.0       339.31      3        10.0               1   
4         3000.0  60.0        67.79      2         1.0               1   
...          ...   ...          ...    ...         ...             ...   
42532     3500.0  36.0       113.39      3         0.0               1   
42533     1000.0  36.0        32.11      2         0.0               1   
42534     2525.0  36.0        80.69      2         0.0               1   
42535     6500.0  36.0       204.84      1         0.0               0   
42550     5000.0  36.0       156.11      1        10.0               2   

       annual_inc  verification_status  loan_status pymnt_plan  ...  \
0         24000.0                    1  

### Dropping the column 'paymt_plan', 'acc_now_delinq' and 'tax_liens' because they have 1, 4 and 1 True values respectively which do not impact the target variable. After dropping, we have 42535 Rows and 28 Columns in the dataset as shown in the output.

In [47]:
print (categorical_handle['pymnt_plan'].value_counts())

False    42534
True         1
Name: pymnt_plan, dtype: int64


In [48]:
drop_col=['pymnt_plan']
categorical_handle = categorical_handle.drop(drop_col,axis=1)

In [49]:
print (categorical_handle['acc_now_delinq'].value_counts())

False    42502
True         4
Name: acc_now_delinq, dtype: int64


In [50]:
drop_col=['acc_now_delinq']
categorical_handle = categorical_handle.drop(drop_col,axis=1)

In [51]:
print (categorical_handle['tax_liens'].value_counts())

False    42429
True         1
Name: tax_liens, dtype: int64


In [52]:
drop_col=['tax_liens']
categorical_handle = categorical_handle.drop(drop_col,axis=1)

In [53]:
print(categorical_handle)

       loan_amnt  term  installment  grade  emp_length  home_ownership  \
0         5000.0  36.0       162.87      2        10.0               1   
1         2500.0  60.0        59.83      3         0.0               1   
2         2400.0  36.0        84.33      3        10.0               1   
3        10000.0  36.0       339.31      3        10.0               1   
4         3000.0  60.0        67.79      2         1.0               1   
...          ...   ...          ...    ...         ...             ...   
42532     3500.0  36.0       113.39      3         0.0               1   
42533     1000.0  36.0        32.11      2         0.0               1   
42534     2525.0  36.0        80.69      2         0.0               1   
42535     6500.0  36.0       204.84      1         0.0               0   
42550     5000.0  36.0       156.11      1        10.0               2   

       annual_inc  verification_status  loan_status  purpose  ... open_acc  \
0         24000.0                

### Dropping the column 'last_credit_pull_d', 'earliest_cr_line' and 'title' because dates, reasons do not impact the target variable. After dropping, we have 42535 Rows and 25 Columns in the dataset as shown in the output.

In [54]:
print (categorical_handle['last_credit_pull_d'].value_counts())

Sep-16    16232
Mar-16      859
Aug-16      771
Apr-16      700
Feb-13      696
          ...  
Feb-08        3
Jul-08        1
Jun-08        1
May-08        1
Jul-07        1
Name: last_credit_pull_d, Length: 111, dtype: int64


In [55]:
drop_col=['last_credit_pull_d']
categorical_handle = categorical_handle.drop(drop_col,axis=1)

In [56]:
print (categorical_handle['earliest_cr_line'].value_counts())

Oct-99    393
Nov-98    390
Oct-00    370
Dec-98    366
Dec-97    348
         ... 
Dec-61      1
Oct-68      1
Aug-67      1
Oct-08      1
Sep-65      1
Name: earliest_cr_line, Length: 530, dtype: int64


In [57]:
drop_col=['earliest_cr_line']
categorical_handle = categorical_handle.drop(drop_col,axis=1)

In [58]:
print (categorical_handle['title'].value_counts())

Debt Consolidation            2259
Debt Consolidation Loan       1760
Personal Loan                  708
Consolidation                  547
debt consolidation             532
                              ... 
shardebt                         1
CAR REPAIRS                      1
Debbie Debt Consolidation        1
Purchase Ford Bronco             1
Pagoda                           1
Name: title, Length: 21253, dtype: int64


In [59]:
drop_col=['title']
categorical_handle = categorical_handle.drop(drop_col,axis=1)

In [60]:
print(categorical_handle)

       loan_amnt  term  installment  grade  emp_length  home_ownership  \
0         5000.0  36.0       162.87      2        10.0               1   
1         2500.0  60.0        59.83      3         0.0               1   
2         2400.0  36.0        84.33      3        10.0               1   
3        10000.0  36.0       339.31      3        10.0               1   
4         3000.0  60.0        67.79      2         1.0               1   
...          ...   ...          ...    ...         ...             ...   
42532     3500.0  36.0       113.39      3         0.0               1   
42533     1000.0  36.0        32.11      2         0.0               1   
42534     2525.0  36.0        80.69      2         0.0               1   
42535     6500.0  36.0       204.84      1         0.0               0   
42550     5000.0  36.0       156.11      1        10.0               2   

       annual_inc  verification_status  loan_status  purpose  ...  \
0         24000.0                    1    

In [61]:
categorical_handle.dtypes

loan_amnt               float64
term                    float64
installment             float64
grade                     int64
emp_length              float64
home_ownership            int64
annual_inc              float64
verification_status       int64
loan_status               int64
purpose                   int64
addr_state                int64
dti                     float64
delinq_2yrs             float64
fico_range_low          float64
fico_range_high         float64
inq_last_6mths          float64
open_acc                float64
pub_rec                 float64
revol_bal               float64
revol_util              float64
total_acc               float64
last_fico_range_high    float64
last_fico_range_low     float64
delinq_amnt             float64
pub_rec_bankruptcies    float64
dtype: object

### Since we removed many columns, we are checking rows with more than half NULL values again. The data set had no such rows. So the number of rows and columns remains the same.

In [62]:
half_count_cols = categorical_handle.shape[1] / 2

In [63]:
print(half_count_cols)

12.5


In [64]:
final = categorical_handle.dropna(thresh=half_count_cols,axis=0)

In [65]:
print(final)

       loan_amnt  term  installment  grade  emp_length  home_ownership  \
0         5000.0  36.0       162.87      2        10.0               1   
1         2500.0  60.0        59.83      3         0.0               1   
2         2400.0  36.0        84.33      3        10.0               1   
3        10000.0  36.0       339.31      3        10.0               1   
4         3000.0  60.0        67.79      2         1.0               1   
...          ...   ...          ...    ...         ...             ...   
42532     3500.0  36.0       113.39      3         0.0               1   
42533     1000.0  36.0        32.11      2         0.0               1   
42534     2525.0  36.0        80.69      2         0.0               1   
42535     6500.0  36.0       204.84      1         0.0               0   
42550     5000.0  36.0       156.11      1        10.0               2   

       annual_inc  verification_status  loan_status  purpose  ...  \
0         24000.0                    1    

### Since we removed many rows, we are checking columns with more than half NULL values again. The data set had no such columns. So the number of rows and columns remains the same.

In [66]:
half_count_rows = final.shape[0] / 2


In [67]:
print(half_count_rows)


21267.5


In [68]:
final = final.dropna(thresh=half_count_rows,axis=1)


In [69]:
print(final)

       loan_amnt  term  installment  grade  emp_length  home_ownership  \
0         5000.0  36.0       162.87      2        10.0               1   
1         2500.0  60.0        59.83      3         0.0               1   
2         2400.0  36.0        84.33      3        10.0               1   
3        10000.0  36.0       339.31      3        10.0               1   
4         3000.0  60.0        67.79      2         1.0               1   
...          ...   ...          ...    ...         ...             ...   
42532     3500.0  36.0       113.39      3         0.0               1   
42533     1000.0  36.0        32.11      2         0.0               1   
42534     2525.0  36.0        80.69      2         0.0               1   
42535     6500.0  36.0       204.84      1         0.0               0   
42550     5000.0  36.0       156.11      1        10.0               2   

       annual_inc  verification_status  loan_status  purpose  ...  \
0         24000.0                    1    

### Finding and Replacing NULL values in 'loan_amnt', 'emp_length' and 'annual_inc' with 0.

In [70]:
final.isnull().sum(axis=0)

loan_amnt                  0
term                       0
installment                0
grade                      0
emp_length              1112
home_ownership             0
annual_inc                 4
verification_status        0
loan_status                0
purpose                    0
addr_state                 0
dti                        0
delinq_2yrs               29
fico_range_low             0
fico_range_high            0
inq_last_6mths            29
open_acc                  29
pub_rec                   29
revol_bal                  0
revol_util                90
total_acc                 29
last_fico_range_high       0
last_fico_range_low        0
delinq_amnt               29
pub_rec_bankruptcies    1365
dtype: int64

In [71]:
final['loan_amnt'].fillna(0, inplace=True)

In [72]:
final['emp_length'].fillna(0, inplace=True)

In [73]:
final['annual_inc'].fillna(0, inplace=True)

### Finding and Replacing NULL values in 'delinq_2yrs', 'inq_last_6mths', 'open_acc', 'pub_rec', 'revol_util', 'total_acc' and 'pub_rec_bankruptcies' with floor of the mean of rest of the column values.

In [74]:
print(final['delinq_2yrs'])

0        0.0
1        0.0
2        0.0
3        0.0
4        0.0
        ... 
42532    NaN
42533    NaN
42534    NaN
42535    NaN
42550    NaN
Name: delinq_2yrs, Length: 42535, dtype: float64


In [75]:
print(final['inq_last_6mths'])

0        1.0
1        5.0
2        2.0
3        1.0
4        0.0
        ... 
42532    NaN
42533    NaN
42534    NaN
42535    NaN
42550    NaN
Name: inq_last_6mths, Length: 42535, dtype: float64


In [76]:
print(final['open_acc'])

0         3.0
1         3.0
2         2.0
3        10.0
4        15.0
         ... 
42532     NaN
42533     NaN
42534     NaN
42535     NaN
42550     NaN
Name: open_acc, Length: 42535, dtype: float64


In [77]:
print(final['pub_rec'])

0        0.0
1        0.0
2        0.0
3        0.0
4        0.0
        ... 
42532    NaN
42533    NaN
42534    NaN
42535    NaN
42550    NaN
Name: pub_rec, Length: 42535, dtype: float64


In [78]:
print(final['revol_util'])

0        0.837
1        0.094
2        0.985
3        0.210
4        0.539
         ...  
42532      NaN
42533      NaN
42534      NaN
42535      NaN
42550      NaN
Name: revol_util, Length: 42535, dtype: float64


In [79]:
print(final['total_acc'])

0         9.0
1         4.0
2        10.0
3        37.0
4        38.0
         ... 
42532     NaN
42533     NaN
42534     NaN
42535     NaN
42550     NaN
Name: total_acc, Length: 42535, dtype: float64


In [80]:
print(final['pub_rec_bankruptcies'])

0        0.0
1        0.0
2        0.0
3        0.0
4        0.0
        ... 
42532    NaN
42533    NaN
42534    NaN
42535    NaN
42550    NaN
Name: pub_rec_bankruptcies, Length: 42535, dtype: float64


In [81]:
print (final['delinq_2yrs'].value_counts())

0.0     37771
1.0      3595
2.0       771
3.0       244
4.0        72
5.0        27
6.0        13
7.0         6
8.0         3
11.0        2
13.0        1
9.0         1
Name: delinq_2yrs, dtype: int64


In [82]:
print (final['inq_last_6mths'].value_counts())

0.0     19657
1.0     11247
2.0      5987
3.0      3182
4.0      1056
5.0       596
6.0       339
7.0       182
8.0       115
9.0        50
10.0       24
12.0       15
11.0       15
15.0        9
14.0        6
13.0        6
18.0        4
16.0        3
24.0        2
17.0        2
19.0        2
28.0        1
33.0        1
27.0        1
31.0        1
25.0        1
32.0        1
20.0        1
Name: inq_last_6mths, dtype: int64


In [83]:
print (final['open_acc'].value_counts())

7.0     4252
8.0     4176
6.0     4172
9.0     3922
10.0    3386
5.0     3368
11.0    2944
4.0     2508
12.0    2398
13.0    2060
3.0     1608
14.0    1597
15.0    1290
16.0    1022
17.0     812
2.0      692
18.0     588
19.0     442
20.0     335
21.0     276
22.0     170
23.0     121
24.0      87
25.0      63
26.0      40
1.0       39
28.0      29
27.0      29
30.0      17
29.0      16
31.0      13
34.0       8
32.0       6
33.0       5
35.0       4
38.0       2
36.0       2
41.0       1
46.0       1
39.0       1
47.0       1
44.0       1
37.0       1
42.0       1
Name: open_acc, dtype: int64


In [84]:
print (final['pub_rec'].value_counts())

0.0    40130
1.0     2298
2.0       64
3.0       11
4.0        2
5.0        1
Name: pub_rec, dtype: int64


In [85]:
print (final['revol_util'].value_counts())

0.0000    1070
0.4070      65
0.0020      64
0.6300      63
0.6660      62
          ... 
0.3329       1
1.0620       1
1.0350       1
0.3019       1
0.2463       1
Name: revol_util, Length: 1119, dtype: int64


In [86]:
print (final['total_acc'].value_counts())

15.0    1552
16.0    1547
17.0    1543
14.0    1531
20.0    1504
        ... 
77.0       1
72.0       1
80.0       1
70.0       1
74.0       1
Name: total_acc, Length: 83, dtype: int64


In [87]:
print (final['pub_rec_bankruptcies'].value_counts())

0.0    39316
1.0     1846
2.0        8
Name: pub_rec_bankruptcies, dtype: int64


In [88]:
import numpy as np
final['delinq_2yrs'].fillna(np.floor(final["delinq_2yrs"].mean()), inplace=True)

In [89]:
print (final['delinq_2yrs'].value_counts())

0.0     37800
1.0      3595
2.0       771
3.0       244
4.0        72
5.0        27
6.0        13
7.0         6
8.0         3
11.0        2
13.0        1
9.0         1
Name: delinq_2yrs, dtype: int64


In [90]:
final['total_acc'].fillna(np.floor(final["total_acc"].mean()), inplace=True)

In [91]:
print (final['total_acc'].value_counts())

15.0    1552
16.0    1547
17.0    1543
14.0    1531
20.0    1504
        ... 
77.0       1
72.0       1
80.0       1
70.0       1
74.0       1
Name: total_acc, Length: 83, dtype: int64


In [92]:
final['revol_util'].fillna(np.floor(final["revol_util"].mean()), inplace=True)

In [93]:
print (final['revol_util'].value_counts())

0.0000    1160
0.4070      65
0.0020      64
0.6300      63
0.6660      62
          ... 
0.3329       1
1.0620       1
1.0350       1
0.3019       1
0.2463       1
Name: revol_util, Length: 1119, dtype: int64


In [94]:
final['pub_rec'].fillna(np.floor(final["pub_rec"].mean()), inplace=True)

In [95]:
print (final['pub_rec'].value_counts())

0.0    40159
1.0     2298
2.0       64
3.0       11
4.0        2
5.0        1
Name: pub_rec, dtype: int64


In [96]:
final['open_acc'].fillna(np.floor(final["open_acc"].mean()), inplace=True)

In [97]:
print (final['open_acc'].value_counts())

7.0     4252
8.0     4176
6.0     4172
9.0     3951
10.0    3386
5.0     3368
11.0    2944
4.0     2508
12.0    2398
13.0    2060
3.0     1608
14.0    1597
15.0    1290
16.0    1022
17.0     812
2.0      692
18.0     588
19.0     442
20.0     335
21.0     276
22.0     170
23.0     121
24.0      87
25.0      63
26.0      40
1.0       39
28.0      29
27.0      29
30.0      17
29.0      16
31.0      13
34.0       8
32.0       6
33.0       5
35.0       4
38.0       2
36.0       2
41.0       1
46.0       1
39.0       1
47.0       1
44.0       1
37.0       1
42.0       1
Name: open_acc, dtype: int64


In [98]:
final['inq_last_6mths'].fillna(np.floor(final["inq_last_6mths"].mean()), inplace=True)

In [99]:
print (final['inq_last_6mths'].value_counts())

0.0     19657
1.0     11276
2.0      5987
3.0      3182
4.0      1056
5.0       596
6.0       339
7.0       182
8.0       115
9.0        50
10.0       24
12.0       15
11.0       15
15.0        9
14.0        6
13.0        6
18.0        4
16.0        3
24.0        2
17.0        2
19.0        2
28.0        1
33.0        1
27.0        1
31.0        1
25.0        1
32.0        1
20.0        1
Name: inq_last_6mths, dtype: int64


In [100]:
final['pub_rec_bankruptcies'].fillna(np.floor(final["pub_rec_bankruptcies"].mean()), inplace=True)

In [101]:
print (final['pub_rec_bankruptcies'].value_counts())

0.0    40681
1.0     1846
2.0        8
Name: pub_rec_bankruptcies, dtype: int64


### Removing the column 'deling_amnt' becuase it has 2 Outliers. After removing we have 42535 Rows and 24 Columns in the dataset.

In [102]:
print(final['delinq_amnt'])

0        0.0
1        0.0
2        0.0
3        0.0
4        0.0
        ... 
42532    NaN
42533    NaN
42534    NaN
42535    NaN
42550    NaN
Name: delinq_amnt, Length: 42535, dtype: float64


In [103]:
print (final['delinq_amnt'].value_counts())

0.0       42504
6053.0        1
27.0          1
Name: delinq_amnt, dtype: int64


In [104]:
drop_col=['delinq_amnt']
final = final.drop(drop_col,axis=1)

# RESULT

### Finally, the data set has 42535 Rows, 24 Columns, no null values and all the columns are of type float or int as shown in the output. 

In [105]:
final.isnull().sum(axis=0)

loan_amnt               0
term                    0
installment             0
grade                   0
emp_length              0
home_ownership          0
annual_inc              0
verification_status     0
loan_status             0
purpose                 0
addr_state              0
dti                     0
delinq_2yrs             0
fico_range_low          0
fico_range_high         0
inq_last_6mths          0
open_acc                0
pub_rec                 0
revol_bal               0
revol_util              0
total_acc               0
last_fico_range_high    0
last_fico_range_low     0
pub_rec_bankruptcies    0
dtype: int64

In [106]:
final.dtypes

loan_amnt               float64
term                    float64
installment             float64
grade                     int64
emp_length              float64
home_ownership            int64
annual_inc              float64
verification_status       int64
loan_status               int64
purpose                   int64
addr_state                int64
dti                     float64
delinq_2yrs             float64
fico_range_low          float64
fico_range_high         float64
inq_last_6mths          float64
open_acc                float64
pub_rec                 float64
revol_bal               float64
revol_util              float64
total_acc               float64
last_fico_range_high    float64
last_fico_range_low     float64
pub_rec_bankruptcies    float64
dtype: object

In [107]:
print(final)

       loan_amnt  term  installment  grade  emp_length  home_ownership  \
0         5000.0  36.0       162.87      2        10.0               1   
1         2500.0  60.0        59.83      3         0.0               1   
2         2400.0  36.0        84.33      3        10.0               1   
3        10000.0  36.0       339.31      3        10.0               1   
4         3000.0  60.0        67.79      2         1.0               1   
...          ...   ...          ...    ...         ...             ...   
42532     3500.0  36.0       113.39      3         0.0               1   
42533     1000.0  36.0        32.11      2         0.0               1   
42534     2525.0  36.0        80.69      2         0.0               1   
42535     6500.0  36.0       204.84      1         0.0               0   
42550     5000.0  36.0       156.11      1        10.0               2   

       annual_inc  verification_status  loan_status  purpose  ...  \
0         24000.0                    1    